clang/lib/3C/ConstraintVariables.cpp

```diff
@@ -112,8 +114,56 @@ PointerVariableConstraint::PointerVariableConstraint(
 
 PointerVariableConstraint::PointerVariableConstraint(DeclaratorDecl *D,
                                                      ProgramInfo &I,
-                                                     const ASTContext &C)
-    : PointerVariableConstraint(D->getType(), D, D->getName(), I, C) {}
+                                                     const ASTContext &C) :
+        PointerVariableConstraint(D->getType(), D, D->getName(),
+                                  I, C) { }
+
+
+// Simple recursive visitor for determining if a type contains a typedef
+// entrypoint is find()
+class TypedefLevelFinder : public RecursiveASTVisitor<TypedefLevelFinder> {
+  public:
+
+    static struct InternalTypedefInfo find(const QualType &QT) {
+      TypedefLevelFinder TLF;
+      QualType tosearch;
+      if (auto TDT = dyn_cast<TypedefType>(QT))
+        tosearch = TDT->desugar();
+      else
+        tosearch = QT;
+      TLF.TraverseType(tosearch);
+      // If we found a typedef the we need to have filled out the name field
+      assert(IMPLIES(TLF.hastypedef, TLF.TDname != ""));
+      struct InternalTypedefInfo info = 
+      	{ TLF.hastypedef, TLF.typedeflevel, TLF.TDname };
+      return info;
+    }
+
+    bool VisitTypedefType(TypedefType *TT) {
+      hastypedef = true;
+      auto TDT = TT->getDecl();
+      TDname = TDT->getNameAsString();
+      return false;
+    }
+
+    bool VisitPointerType(PointerType *PT) {
+      typedeflevel++;
+      return true;
+    }
+
+    bool VisitArrayType(ArrayType *AT) {
+      typedeflevel++;
+      return true;
+    }
```

Comment:
Same question here that I had above: the assumption seems to be that your type is a linear sequence of pointers, but for a function pointer type it may not be.